compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala

```diff
@@ -309,26 +311,48 @@ trait ConstraintHandling[AbstractContext] {
    * At this point we also drop the @Repeated annotation to avoid inferring type arguments with it,
    * as those could leak the annotation to users (see run/inferred-repeated-result).
    */
-  def widenInferred(inst: Type, bound: Type)(implicit actx: AbstractContext): Type = {
-    def widenOr(tp: Type) = {
+  def widenInferred(inst: Type, bound: Type)(implicit actx: AbstractContext): Type =
+
+    def isRestricted(tp: Type) = tp.typeSymbol == defn.EnumValueClass // for now, to be generalized later
+
+    def dropRestricted(tp: Type): Type = tp.dealias match
+      case tpd @ AndType(tp1, tp2) =>
+        if isRestricted(tp1) then tp2
+        else if isRestricted(tp2) then tp1
+        else
+          val tpw = tp.derivedAndType(dropRestricted(tp1), dropRestricted(tp2))
+          if tpw ne tpd then tpw else tp
+      case _ =>
+        tp
+
+    def widenRestricted(tp: Type) =
+      val tpw = dropRestricted(tp)
+      if (tpw ne tp) && (tpw <:< bound) then tpw else tp
+
+    def widenOr(tp: Type) =
       val tpw = tp.widenUnion
       if (tpw ne tp) && (tpw <:< bound) then tpw else tp
-    }
-    def widenSingle(tp: Type) = {
+
+    def widenSingle(tp: Type) =
       val tpw = tp.widenSingletons
       if (tpw ne tp) && (tpw <:< bound) then tpw else tp
-    }
+
     def isSingleton(tp: Type): Boolean = tp match
       case WildcardType(optBounds) => optBounds.exists && isSingleton(optBounds.bounds.hi)
       case _ => isSubTypeWhenFrozen(tp, defn.SingletonType)
+
     val wideInst =
-      if isSingleton(bound) then inst else widenOr(widenSingle(inst))
+      if isSingleton(bound) then inst
+      else /*widenRestricted*/(widenOr(widenSingle(inst)))
+        // widenRestricted is currently not called since it's special cased in `dropEnumValue`
+        // in `Namer`. It's left in here in case we want to generalize the scheme to other
+        // "protected inheritance" classes.
```

Comment:
I'm not sure we should leave dead code like this around without a clear idea of what we want to do with it. Is this something you're planning to experiment with more? I think it would make sense to use this mechanism for getting rid of Serializable and Product so that they don't show up when lubbing case classes.

Reply:
See #9028.

Reply:
Very nice, thanks! Can you leave a reference to this issue number in the comment?